modules/test/portlet-container-test/test/integration/com/liferay/portlet/container/test/BaseTestCase.java

```diff
@@ -0,0 +1,243 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+import java.util.Arrays;
+import java.util.Dictionary;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.util.tracker.ServiceTracker;
+
+import org.springframework.mock.web.MockHttpServletRequest;
+
+/**
+ * @author Raymond Augé
+ */
+public class BaseTestCase {
+
+	@Before
+	public void setUp() throws Exception {
+		group = GroupTestUtil.addGroup();
+
+		layout = LayoutTestUtil.addLayout(group);
+
+		PermissionChecker permissionChecker =
+			PermissionCheckerFactoryUtil.create(TestPropsValues.getUser());
+
+		PermissionThreadLocal.setPermissionChecker(permissionChecker);
+	}
+
+	@After
+	public void tearDown() throws Exception {
+		PermissionThreadLocal.setPermissionChecker(null);
+
+		if (serviceTracker != null) {
+			serviceTracker.close();
+		}
+
+		for (ServiceRegistration<?> serviceRegistration :
+				serviceRegistrations) {
+
+			serviceRegistration.unregister();
+		}
+
+		serviceRegistrations.clear();
+	}
+
+	protected String drain(InputStream stream) throws IOException {
+		if (stream == null) {
+			return "";
+		}
+
+		byte[] bytes = new byte[100];
+		StringBuffer buffer = new StringBuffer(500);
+		int length = 0;
+
+		while ((length = stream.read(bytes)) != -1) {
+			String chunk = new String(bytes, 0, length);
+			buffer.append(chunk);
+		}
+
+		return buffer.toString();
+	}
+
+	protected BundleContext getBundleContext() {
+		Bundle bundle = FrameworkUtil.getBundle(getClass());
+
+		return bundle.getBundleContext();
+	}
+
+	protected MockHttpServletRequest getRequest() throws Exception {
+		MockHttpServletRequest request = new MockHttpServletRequest();
+
+		ThemeDisplay themeDisplay = ThemeDisplayFactory.create();
+
+		Company company = CompanyLocalServiceUtil.getCompany(
+			layout.getCompanyId());
+
+		themeDisplay.setCompany(company);
+		themeDisplay.setLayout(layout);
+		themeDisplay.setPlid(layout.getPlid());
+		themeDisplay.setPortalURL(TestPropsValues.PORTAL_URL);
+		themeDisplay.setRequest(request);
+		themeDisplay.setScopeGroupId(group.getGroupId());
+		themeDisplay.setSiteGroupId(group.getGroupId());
+		themeDisplay.setUser(TestPropsValues.getUser());
+
+		request.setAttribute(WebKeys.LAYOUT, layout);
+		request.setAttribute(WebKeys.THEME_DISPLAY, themeDisplay);
+
+		return request;
+	}
+
+	protected Map<String, List<String>> request(String urlSpec)
+		throws IOException {
+
+		return request(urlSpec, null);
+	}
+
+	protected Map<String, List<String>> request(
+			String urlSpec, Map<String, List<String>> headers)
+		throws IOException {
+
+		URL url = new URL(urlSpec);
```

Comment:
I would avoid doing any real network based communication inside this test. This reminds me the ComboServlet issues we had on CI, same problem could happen. We should able to get this done via RequestDispatcher.
The worst case, we can use a URLStreamHandlerFactory to cut off the real network communication, to do things directly.